orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy

```diff
@@ -850,6 +870,113 @@ class TaskController {
+   *
+   *<p>
+   *   2. It then uses the list of pipeline config ids obtained from step 1 and gets all the valid executions
+   *   associated with each one of them. The valid executions are found after applying the execution criteria.
+   *
+   *<p>
+   *   3. It then processes n pipeline executions at a time to retrieve the complete execution details. In addition,
+   *      we make use of a configured thread pool to process multiple batches of n executions in parallel.
+   */
+  private List<PipelineExecution> optimizedGetPipelineExecutions(String application,
+  List<String> front50PipelineConfigIds, ExecutionCriteria executionCriteria) {
+    List<PipelineExecution> finalResult = []
+    log.info("running optimized execution retrieval process with: " +
+        "${this.configurationProperties.getMaxExecutionRetrievalThreads()} threads and processing" +
+        " ${this.configurationProperties.getMaxNumberOfPipelineExecutionsToProcess()} pipeline executions at a time")
+
+    List<String> commonPipelineConfigIdsInFront50AndOrca
+    try {
+      List<String> allOrcaPipelineConfigIds = executionRepository.retrievePipelineConfigIdsForApplication(application)
+      log.info("found ${allOrcaPipelineConfigIds.size()} pipeline config ids for application: $application in orca")
+      commonPipelineConfigIdsInFront50AndOrca = front50PipelineConfigIds.intersect(allOrcaPipelineConfigIds)
+      log.info("found ${commonPipelineConfigIdsInFront50AndOrca.size()} pipeline config ids that are common in orca " +
+          "and front50 for application: $application. " +
+          "Saved ${front50PipelineConfigIds.size() - commonPipelineConfigIdsInFront50AndOrca.size()} extra pipeline " +
+          "config id queries")
+    } catch (Exception e) {
+      log.warn("retrieving pipeline config ids from orca db failed. using the result obtained from front50 ", e)
+      commonPipelineConfigIdsInFront50AndOrca = front50PipelineConfigIds
+    }
+
+    if (commonPipelineConfigIdsInFront50AndOrca.size() == 0 ) {
+      log.info("no pipeline config ids found.")
+      return finalResult
+    }
+
+    // get complete list of executions based on the execution criteria
+    log.info("filtering pipeline executions based on the execution criteria: " +
+        "limit: ${executionCriteria.getPageSize()}, statuses: ${executionCriteria.getStatuses()}")
+    List<String> filteredPipelineExecutionIds = executionRepository.retrieveAndFilterPipelineExecutionIdsForApplication(
+        application,
+        commonPipelineConfigIdsInFront50AndOrca,
+        executionCriteria
+    )
+    if (filteredPipelineExecutionIds.size() == 0) {
+      log.info("no pipeline executions found")
+      return finalResult
+    }
+
+    // need to define a new executor service since we want a dedicated set of threads to be made available for every
+    // new request for performance reasons
+    ExecutorService executorService = Executors.newFixedThreadPool(
```

Comment:
It's not quite a thread-per-request since the number of threads is set by a config prop (default 10).

https://github.com/spinnaker/orca/pull/4803 comes next, with more to follow as we move requests to a read-only database replica.

Reply:
Yeah saw the properties but generating a thread pool on the request even WITH limits is still backed by a queue ... It's still shutdown but this could absolutely hammer a database badly because of a potential exponential call graph. 
 E.g. a browser can repeat calls which then would make new calls on timeouts that'd overload this badly. The RO PR should help A LOT with that however, and IN THEORY this should improve performance.  It's just a concern based upon how I've seen these API's get called.

Reply:
Yup.  We also ended up making changes in deck to improve things.  Those are gonna be harder for us to upstream, but we'll try.